Fixes for interactive database changes.

johnworth opened the pull request:
There's now a foreign key to the interactive_apps_proxy_settings table
in the container_settings table.

The foreign key into the container_settings table was dropped from the
interactive_apps_proxy_settings table.

A uniqueness constraint was added to the
interactive_apps_proxy_settings.id column.

The SQL files for setting up the various tables and constraints were
modified to account for the above.

A new version entry was added to the database.

johnworth committed:
Fixed a bunch of interapps related issues.

There's now a foreign key to the interactive_apps_proxy_settings table
in the container_settings table.

The foreign key into the container_settings table was dropped from the
interactive_apps_proxy_settings table.

A uniqueness constraint was added to the
interactive_apps_proxy_settings.id column.

The SQL files for setting up the various tables and constraints were
modified to account for the above.

A new version entry was added to the database.

johnworth committed:
Merge branch 'master' into interapps

slr71 reviewed (approved):
LGTM :shipit:

psarando commented on src/main/constraints/00_71_container_settings_fkey.sql:
```diff
@@ -0,0 +1,6 @@
+
+ALTER TABLE ONLY container_settings
+    ADD CONSTRAINT container_settings_interactive_apps_proxy_settings_id_fkey
+    FOREIGN KEY(interactive_apps_proxy_settings_id)
```
So far we've been putting `FOREIGN KEY` constraints in the regular `src/main/constraints/XX_table_name.sql` files, and only primary keys (since they have to be defined before other constraints) have been going into `constraints/00_XX_table_name_pkey.sql` files.

Maybe this should just go into the `constraints/71_container_settings.sql` file for consistency?

psarando commented on src/main/conversions/v2.21.0/c2_21_0_2018050101.clj:
```diff
@@ -0,0 +1,44 @@
+  []
+  (exec-sql-statement
+   "ALTER TABLE ONLY interactive_apps_proxy_settings
+    DROP COLUMN IF EXISTS container_settings_id"))
```
Since `tables/84_interapps_proxy_settings.sql` (added by #23) isn't in prod yet, we don't need to alter it in this conversion; because we'll import prod's data (without this table), and the `2.21.0:20180426.01` conversion will add this table with the new column already in `tables/84_interapps_proxy_settings.sql` (added by this PR).

Same for the dropped constraint in `constraints/84_interapps_proxy_settings.sql` (it doesn't need to be removed in this conversion).

psarando replied:
For that matter, we don't even need a new conversion either. The new `container_settings` column and constraint can be safely added to the previous conversion.

slr71 replied:
My feeling on this is that it doesn't hurt to have this conversion act as though the intent of the previous conversion happened. Attempting to drop the columns with these statements won't do any harm, and it does ensure that this conversion will succeed if it's ever applied to a database that was converted to version `2.21.0:201804226.01`.

psarando commented on src/main/conversions/v2.21.0/c2_21_0_2018050101.clj:
```diff
@@ -0,0 +1,44 @@
+  []
+  (exec-sql-statement
+   "ALTER TABLE ONLY interactive_apps_proxy_settings
+    ADD CONSTRAINT interactive_apps_proxy_settings_id_unique UNIQUE (id)"))
```
I thought the `UNIQUE` constraint was automatically added to primary keys?

johnworth replied:
When I tested creating the database from scratch it was barking at me about the missing uniqueness constraint.

psarando replied:
Ah, that's probably because `00_71_container_settings_fkey.sql` is loaded first, before loading `84_interapps_proxy_settings.sql` which defines the primary key (which would make it unique). If the constraint in `00_71_container_settings_fkey.sql` is moved to `71_container_settings.sql`, and `84_interapps_proxy_settings.sql` is renamed to `00_84_interapps_proxy_settings_pkey.sql`, then these files will be loaded in the correct order.

psarando commented on src/main/constraints/84_interapps_proxy_settings.sql:
```diff
@@ -3,8 +3,3 @@ SET search_path = public, pg_catalog;
 ALTER TABLE ONLY interactive_apps_proxy_settings
     ADD CONSTRAINT interactive_apps_proxy_settings_id_pkey
     PRIMARY KEY(id);
```
As mentioned in the comment above, primary key constraints have been going into `constraints/00_XX_table_name_pkey.sql` files. So we should probably rename this file to `00_84_ interactive_apps_proxy_settings_pkey.sql`, since it only contains a `PRIMARY KEY` constraint now.

psarando commented on src/main/conversions/v2.21.0/c2_21_0_2018050101.clj:
```diff
@@ -0,0 +1,44 @@
+
+(defn- add-constraint
+  []
+  (load-sql-file "constraints/00_71_container_settings_fkey.sql"))
```
If we move the `container_settings_interactive_apps_proxy_settings_id_fkey` constraint into the existing `constraints/71_container_settings.sql` file, then we'll have to update this function to manually add the new constraint, since reloading that file will probably cause an error trying to re-add the existing `container_settings_tools_id_fkey` constraint.

psarando reviewed (changes requested):
I think we could cleanup this conversion for consistency and to remove redundant conversions.

johnworth committed:
Moved container_settings and interapps constraints to match conventions

johnworth committed:
Moved more logic around to match conventions.

Moved the container_ports pkey constraint into
00_83_container_ports.sql.

Modified the c2_21_0_2018042601.clj conversion to use the new layout.

Removed the unneeded conversion.

psarando commented on src/main/data/99_version.sql:
```diff
@@ -93,3 +93,4 @@ INSERT INTO version (version) VALUES ('2.18.0:20171113.01');
 INSERT INTO version (version) VALUES ('2.20.0:20180326.01');
 INSERT INTO version (version) VALUES ('2.21.0:20180419.01');
 INSERT INTO version (version) VALUES ('2.21.0:20180426.01');
+INSERT INTO version (version) VALUES ('2.21.0:20180501.01');
```
Look like we don't need `2.21.0:20180501.01` in `99_version.sql` anymore.

psarando reviewed (approved):
Revert changes to `99_version.sql` and this PR LGTM 👍

johnworth committed:
Removed unneeded version from 99_version.sql
